### PR TITLE
Final tests

### PR DESCRIPTION
**Pull request by Lobotuerk**
Signed-off-by: lobotuerk <jtlorente@ekumenlabs.com>
Full matrix3 Tests

### CODE REVIEW

File: course/test/src/matrix3_TEST.cc [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+namespace ekumen {
+namespace math {
+namespace test {
+namespace {
+
+  GTEST_TEST(Matrix3Test, Matrix3FullTests) {
+    const double kTolerance{1e-12};
+    Matrix3 m1{{1., 2., 3.}, {4., 5., 6.}, {7., 8., 9.}};
+    const Matrix3 m2{1., 2., 3., 4., 5., 6., 7., 8., 9.};
+    const Matrix3 m3{1., 2., 3., 4., 5., 6., 7., 8., 9.};
```
Comment (glpuga) [inline course/test/src/matrix3_TEST.cc:25]: Also please create and test with the default constructor and test that all its elements are zero. Also reimplement the default constructor to work as a delegating constructor, for practice.

Comment (glpuga) [inline course/test/src/matrix3_TEST.cc:25]: Make the Vector3 class movable, and follow the rule of 5 both in Vector3 and in Matrix3.

For the sake of the exercise make `Vector3` something that moving from is worth it. I suggest instead of storing the elements as "double x_,y_,z_;" store them in a struct:
```
struct Elements {
  double x_,y_,z_;
};
```
then in `Vector3`, declare the storage as:
```
private:
   Elements *v_;
```
such that you allocate the space for `v_` when you construct with `new` and deallocate it with `delete` in the destructor (if needed, beware when you move from an object).

The rest of the changes should be trivial, `x_` will become `v_->x`, `y_` will become`v_->y` and `z_` will become `v_->z`, and your test coverage will protect you against messing up during refactoring.

Again, make sure you follow the rule of five in `Vector3`, and then do the same in `Matrix3`. If you do it right, you'll see  no differences... if you mess the move copy operations you may have memory access problems and premature death of the test.

Notice you should add test cases for your move operations , to make sure your move operations are getting selected when you think they are; there are no hard rules for how to test that, but in your case you can add a "bool moved()" method that returns `true` if the Vector3 object is the result of a move operation/constructor, and `false` in other cases. You can then test its value in the resulting object after a copy/move test.

Comment (glpuga) [reply course/test/src/matrix3_TEST.cc:25]: Keep this in mind: https://stackoverflow.com/questions/7027523/what-can-i-do-with-a-moved-from-object

Comment (glpuga) [reply course/test/src/matrix3_TEST.cc:25]: Never mind, you're creating a default object below.

File: course/test/src/matrix3_TEST.cc [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+    const std::vector<Vector3> kExpectedRows{Vector3(1., 2., 3.),
+        Vector3(4., 5., 6.), Vector3(7., 8., 9.)};
+    const std::vector<Vector3> kExpectedCols{Vector3(1., 4., 7.),
+        Vector3(2., 5., 8.), Vector3(3., 6., 9.)};
+
+    for (const Vector3& r : kExpectedRows) {
+      int found{0};
+      for (int i = 0; i < 3; ++i) {
+        if (r == m2.row(i)) {
```
Comment (glpuga) [inline course/test/src/matrix3_TEST.cc]: Isn't it enough to test that `m2.row(0) == kExpectedRows[0]` and so and so? Here we are testing that we find the vectors we expected, but not necessarily in the order we think they should be.

File: course/test/src/matrix3_TEST.cc [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+          found++;
+          break;
+        }
+      }
+      ASSERT_EQ(found, 1);
+    }
+
+    for (const Vector3& c : kExpectedCols) {
```
Comment (glpuga) [inline course/test/src/matrix3_TEST.cc]: Some comment as for the loop above.

**Review by glpuga — CHANGES_REQUESTED** [review]
Review done. Sorry for the delay!

File: course/test/src/matrix3_TEST.cc (hunk 2 above)
Comment (Lobotuerk) [reply course/test/src/matrix3_TEST.cc]: Agree

File: course/test/src/matrix3_TEST.cc (hunk 3 above)
Comment (Lobotuerk) [reply course/test/src/matrix3_TEST.cc]: Same as the other one

File: course/test/src/matrix3_TEST.cc (hunk 1 above)
Comment (Lobotuerk) [reply course/test/src/matrix3_TEST.cc:25]: I cant make it a delegating constructor, cppcheck does not like it `(performance) Variable 'row1_' is assigned in constructor body. Consider performing initialization in initialization list.`

Comment (glpuga) [reply course/test/src/matrix3_TEST.cc:25]: haha, that's fine. Let's follow the static analyzer advice.

Comment (Lobotuerk) [reply course/test/src/matrix3_TEST.cc:25]: > The rest of the changes should be trivial, x_ will become v_->x, y_ will becomev_->y and z_ will become v_->z, and your test coverage will protect you against messing up during refactoring.

Im having some issues with this Gera, all the lines are giving me 
` error: 'v_' was not declared in this scope
     v_->x_ *= obj;` D:

Comment (glpuga) [reply course/test/src/matrix3_TEST.cc:25]: > Im having some issues with this Gera, all the lines are giving me

It builds for me, are you still having issues?

Comment (Lobotuerk) [reply course/test/src/matrix3_TEST.cc:25]: @glpuga Wut, after pulling?

**Review by glpuga — APPROVED** [review]
Awesome work! Please, do the honors and merge away!
